lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,559 @@
+        with_log(caller, "initialize node", fn ->
+          initialize_node(child_node)
+        end)
+
+      send(primary_pid, :node_initialized)
+
+      runtime = %{runtime | node: child_node, server_pid: server_pid, pod_name: pod_name}
+      send(caller, {:runtime_connect_done, self(), {:ok, runtime}})
+
+      {:noreply, %{state | primary_ref: primary_ref}}
+    else
+      {:error, error} ->
+        send(caller, {:runtime_connect_done, self(), {:error, error}})
+
+        {:stop, :shutdown, state}
+    end
+  end
+
+  @impl true
+  def handle_info({:DOWN, ref, :process, _pid, _reason}, state) when ref == state.primary_ref do
+    {:stop, :shutdown, state}
+  end
+
+  def handle_info({port, _message}, state) when is_port(port) do
+    {:noreply, state}
+  end
+
+  defp get_free_port!() do
+    {:ok, socket} = :gen_tcp.listen(0, active: false, reuseaddr: true)
+    {:ok, port} = :inet.port(socket)
+    :gen_tcp.close(socket)
+    port
+  end
+
+  defp with_log(caller, name, fun) do
+    send(caller, {:runtime_connect_info, self(), name})
+
+    {microseconds, result} = :timer.tc(fun)
+    milliseconds = div(microseconds, 1000)
+
+    case result do
+      {:error, error} ->
+        Logger.debug("[K8s runtime] #{name} FAILED in #{milliseconds}ms, error: #{error}")
+
+      _ ->
+        Logger.debug("[K8s runtime] #{name} finished in #{milliseconds}ms")
+    end
+
+    result
+  end
+
+  defp with_pod_events(caller, name, req, namespace, pod_name, fun) do
+    with_log(caller, name, fn ->
+      runtime_pid = self()
+
+      event_watcher_pid =
+        spawn_link(fn ->
+          watch_result =
+            req
+            |> Req.merge(
+              resource_path: "api/v1/namespaces/:namespace/events/:name",
+              resource_list_path: "api/v1/namespaces/:namespace/events"
+            )
+            |> Kubereq.watch(namespace,
+              field_selectors: [
+                {"involvedObject.kind", "Pod"},
+                {"involvedObject.name", pod_name}
+              ]
+            )
+
+          case watch_result do
+            {:ok, stream} ->
+              Enum.each(stream, fn event ->
+                Logger.debug(~s'[K8s runtime] Pod event: "#{event["object"]["message"]}"')
+                send(caller, {:runtime_connect_info, runtime_pid, event["object"]["message"]})
+              end)
+
+            _error ->
+              :ok
+          end
+        end)
+
+      result = fun.()
+      Process.exit(event_watcher_pid, :normal)
+      result
+    end)
+  end
+
+  defp create_pod(req, config, runtime_data, remote_port) do
+    %{
+      pod_template: pod_template,
+      docker_tag: docker_tag,
+      home_pvc: home_pvc,
+      namespace: namespace
+    } = config
+
+    manifest =
+      pod_template.template
+      |> Pod.pod_from_template()
+      |> Pod.add_env_vars([
+        %{"name" => "LIVEBOOK_RUNTIME", "value" => runtime_data},
+        %{"name" => "ERL_AFLAGS", "value" => "-proto_dist inet6_tcp"},
```

Comment:
Is it advised to use ipv6 with k8s?

Reply:
ah no... probably copy-paste. I will remove